Gd 189

Barmalej24 opened the pull request.

Barmalej24 committed:
Added logic for updating message sending status

Barmalej24 committed:
Added dockerfile

Barmalej24 committed:
Added Docker

Barmalej24 committed:
fixed name

Barmalej24 committed:
fixed name 2

Barmalej24 committed:
fixed Exception logic SenderService
fixed ask queue EmailConsumerService

oleggolen commented on docker-compose.yml:
```diff
@@ -0,0 +1,29 @@
+    image: postgres:latest
+    container_name: local_pgdb_email
+    restart: always
+    environment:
+      POSTGRES_DB: ${POSTGRES_DB}
+      POSTGRES_USER: ${POSTGRES_USER}
+      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
```
Лучше вынести в .env файл и добавить его в ingore и при CI в будущем подставлять

Barmalej24 replied:
параметры там и хранятся, и он в игноре. Или другая какая-то схема имелась введу?

oleggolen replied:
Да вообще весь этот блок убрать и использовать env_file, который будет подставляться из CI или быть локально в игноре https://docs.docker.com/compose/how-tos/environment-variables/set-environment-variables/ вот тут можно посмотреть про env_file и примерчик примерно вот тут https://www.cherryservers.com/blog/set-docker-environment-variables и тогда будет красиво

Barmalej24 replied:
понял. сам файл указать. исправлю.

oleggolen commented on docker-compose.yml:
```diff
@@ -0,0 +1,29 @@
+services:
+  backend:
+    container_name: emailsender_api
+    build:
+     context: .
+     dockerfile: EmailSenderMicroservice/Dockerfile
+    restart: always
+    ports:
+      - "3001:8080"
+    depends_on:
+      - database
```
Добавить health-chacks и самому сервису

Barmalej24 replied:
Добавил

oleggolen commented on EmailSenderMicroservice.Domain/Entities/Message.cs:
Поле MessageType можно наверное переименовать, как MessageTemplateName или комментарий поменять, а то уж сильно и не коррелируется, и странно, что она просто строка, если следовать комментарию

Barmalej24 replied:
Исправил описание

oleggolen commented on EmailSenderMicroservice.DataAccess/Repositories/MessageRepository.cs:
В AddAsync тоже надо передавать cancellationToken, а не только на сохранение изменений

Barmalej24 replied:
добавил

oleggolen reviewed (dismissed):
Совсем уже мелкие исправления и просто придирки. Шикарный код, так держать, надеюсь дальше так и будет

Barmalej24 committed:
Added healthcheck for API
Added cancellationToken for AddAsync's functions
Chenged discription MessageType

m-a-ilnitsky commented on EmailSenderMicroservice.Application/Services/SenderService.cs:
```diff
@@ -25,16 +26,17 @@ public class SenderService(ISettingService settingService, IMessageService messa
         /// <param name="isHtml">Указывает, является ли текст письма HTML-контентом.</param>
         /// <returns>Асинхронная задача.</returns>
```
Здесь по идее должно быть не про Task, и даже не про Task<bool>, а про сам bool

Barmalej24 replied:
обновил описание.

m-a-ilnitsky commented on EmailSenderMicroservice.Domain/ValueObjects/Abstraction/StringValueObject.cs:
```diff
@@ -0,0 +1,77 @@
+        /// Проверяет, соответствует ли переданная строка формату корректного значения объекта.
+        /// </summary>
+        /// <param name="value">Строка для проверки.</param>
+        protected abstract void IsValid(string value);
```
Сейчас по имени кажется что это метод который в случае обнаружения ошибки вернёт false.
Но это void-метод, поэтому такое имя не подходит.
И т.к. это void-метод, логично ожидать, что это метод, который в случае обнаружения ошибки бросит исключение.
Тогда такой метод правильнее назвать CheckValue или ValidateValue, или просто Validate

Barmalej24 replied:
переименован Validate

m-a-ilnitsky commented on EmailSenderMicroservice.Domain/ValueObjects/Abstraction/StringValueObject.cs:
```diff
@@ -0,0 +1,77 @@
+        /// <returns><c>true</c>, если оба объекта равны; в противном случае <c>false</c>.</returns>
+        public static bool operator ==(StringValueObject left, StringValueObject right)
+        {
+            if (ReferenceEquals(left, null))
+                return ReferenceEquals(right, null);
+            return left.Equals(right);
```
По идее этот код можно сократить до одной строки со стрелочным синтаксисом, если использовать статический Equals(left, right), который в итоге сам под капотом вызовет нестатический left.Equals(right) после всех обработок null

Barmalej24 replied:
использовал.

m-a-ilnitsky commented on EmailSenderMicroservice.Domain/ValueObjects/Email.cs:
```diff
@@ -1,94 +1,31 @@
+        protected override void IsValid(string value)
         {
-            if (!IsValid(value))
+            if (!(!string.IsNullOrWhiteSpace(value) && ValidationRegex.IsMatch(value)))
```
Лучше раскрывать отрицания от скобки, тогда вычисление результата выражения будет более эффективным, т.к. для получения правильного результата не всегда будет нужно вычислять всё логическое выражение целиком

Barmalej24 replied:
(string.IsNullOrWhiteSpace(value) || !ValidationRegex.IsMatch(value))

m-a-ilnitsky reviewed (approved):
На мой взгляд, всё достаточно хорошо

oleggolen commented on EmailSenderMicroservice/Program.cs:
```diff
@@ -70,10 +79,18 @@ public static void Main(string[] args)
+
+            app.UseCors();
+
+            app.MapHealthChecks("health", new HealthCheckOptions
+            {
+                ResponseWriter = UIResponseWriter.WriteHealthCheckUIResponse
+            });
```
Лучше сделать составной helthcheck, чтобы проверять то, что и соединение с базой данной установилось, и контейнер поднялся и всё такое, то есть что приложение полностью работоспособно, а не только загрузилось

Barmalej24 replied:
Это вроде все есть 69 строка
 builder.Services.AddHealthChecks()
     .AddNpgSql(connectionString)
     .AddDbContextCheck<EmailSenderMicroserviceDbContext>();

Barmalej24 replied:
добавил .AddRabbitMQ()

Barmalej24 replied:
{
  "status": "Healthy",
  "totalDuration": "00:00:00.6962689",
  "entries": {
    "npgsql": {
      "data": {

      },
      "duration": "00:00:00.0144494",
      "status": "Healthy",
      "tags": []
    },
    "rabbitmq": {
      "data": {

      },
      "duration": "00:00:00.6914763",
      "status": "Healthy",
      "tags": []
    },
    "EmailSenderMicroserviceDbContext": {
      "data": {

      },
      "duration": "00:00:00.0472624",
      "status": "Healthy",
      "tags": []
    }
  }
}

oleggolen reviewed (approved):
Ну вот теперь всё тут отлично, добавить более составной helthcheck для приложения, что и БД и брокер и всё прочие установилось, и тогда полностью готовое приложение куда угодно, ну а также спрямтать переменные, как нужно

ConstantinProg reviewed (approved):
Всё очень хорошо. Давайте двигаться дальше.
